### ISSUE TITLE
Get nan loss using RiemannianAdam

### ISSUE THREAD

**Issue by git4sun** [geoopt/geoopt#227]
**Describe the bug**
I was using RiemannianAdam to optimize a simple linear network for poincare embedding. The input in a vector of dimension (1024) and the output dimension is (20). But the RiemannianAdam optimization step result in nan loss. Tried RiemannianSGD directly, model will learn but learning rate needs to be adjusted. 

**To Reproduce**
```
class PoincareModel(torch.nn.Module):
    def __init__(self, input_dim, output_dim, init_weights=1e-3, epsilon=1e-7):
        super().__init__()
        self.embedding = nn.Linear(input_dim, output_dim)
        self.embedding.weight.data.uniform_(-init_weights, init_weights)
        self.epsilon = epsilon

    def dist(self, u, v):
        sqdist = torch.sum((u - v) ** 2, dim=-1)
        squnorm = torch.sum(u ** 2, dim=-1)
        sqvnorm = torch.sum(v ** 2, dim=-1)
        x = 1 + 2 * sqdist / ((1 - squnorm) * (1 - sqvnorm)) + self.epsilon
        z = torch.sqrt(x ** 2 - 1)
        return torch.log(x + z)

    def forward(self, inputs):
        e = self.embedding(inputs)
        o = e.narrow(dim=1, start=1, length=e.size(1) - 1)
        s = e.narrow(dim=1, start=0, length=1).expand_as(o)

        return self.dist(s, o)
```

Running on torch 2.4.0, macos, python 3.12.4.

Any idea what could be the problem?

**Comment by MostlyKIGuess** [comment]
@git4sun Were you able to resolve this?

**Comment by git4sun** [comment]
no, gave up.

**Comment by SauravMaheshkar** [comment]
On some deeper thought, I think the problem is in the distance formula you have in your model

```python
1 + 2 * sqdist / ((1 - squnorm) * (1 - sqvnorm)) + self.epsilon
```

This becomes negative when either vector has norm > 1 (outside Poincare ball), causing `sqrt(x ** 2 - 1)` to produce NaN.

We could add a projection at the start of each optimizer step to ensure stability here maybe.
